### PR TITLE
Proto Organizations

### PR DESCRIPTION
**Pull request by ryanpbrewster**
A proto-Organization is a placeholder organization that holds all the
information necessary to create a real organization (e.g., the owner, name,
and a list of members to invite). There are two db tables for proto orgs,
one for the proto-orgs themselves and another for all the members associated
with them.

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
@leogrim, I opted to keep the name `ProtoOrganization` because `OrganizationPrototype` is, in my opinion, too close to actual classes like `OrganzationMembership` and `OrganizationInvite`. I want anything that starts with `Organization` to be associated with real orgs.

**Comment by leogrim** [comment]
It is meant to become a real `Organization` though, so it makes sense to me to keep the name scheme in line with that ecosystem, don't you think?

**Comment by leogrim** [comment]
I don't feel strongly about it actually, `ProtoOrganization` just feels hard to say :)
More importantly, we should discuss tracking continuity from a proto-org to and org, primary ids will be different so we may want to introduce a stable tracking identifier. What do you think?

**Comment by ryanpbrewster** [comment]
What if we assign the proto-org a handle?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/ProtoOrganizationCommander.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  def addMembers(protoOrgId: Id[ProtoOrganization], userIds: Set[Id[User]]): Future[Unit] = SafeFuture {
+    db.readWrite { implicit session =>
+      val existingMemberships = protoOrgMembershipRepo.getAllByProtoOrganization(protoOrgId, states = ProtoOrganizationMembershipStates.all).filter {
+        _.userId.nonEmpty
+      }.toSet
+
+      val inactiveMemberships = existingMemberships.filter(_.state == ProtoOrganizationMembershipStates.INACTIVE)
+      inactiveMemberships.foreach { m => protoOrgMembershipRepo.save(m.withState(ProtoOrganizationMembershipStates.ACTIVE)) }
+
+      val newUserIds = userIds -- existingMemberships.filter(_.userId.isDefined).map(_.userId.get)
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/ProtoOrganizationCommander.scala]: `existingMemberships.flatMap{_.userId}` ?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/ProtoOrganizationCommander.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+      newUserIds.foreach { uid =>
+        protoOrgMembershipRepo.save(ProtoOrganizationMembership(protoOrgId = protoOrgId, userId = Some(uid)))
+      }
+    }
+  }
+  def removeMembers(protoOrgId: Id[ProtoOrganization], userIds: Set[Id[User]]): Future[Unit] = SafeFuture {
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/ProtoOrganizationCommander.scala]: this is not yet in the trait.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/ProtoOrganizationCommander.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    db.readWrite { implicit session =>
+      val existingMemberships = protoOrgMembershipRepo.getAllByProtoOrganization(protoOrgId, states = ProtoOrganizationMembershipStates.all).filter {
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/ProtoOrganizationCommander.scala]: you can just retrieve active members, instead of `all`.

**Pull request merged**